backend/aim/metrics/m17/m17_distinct_values_of_hue_saturation_and_value.py

```diff
@@ -0,0 +1,159 @@
+from PIL import Image
+from pydantic import HttpUrl
+
+# First-party modules
+from aim.common.constants import GUI_TYPE_DESKTOP
+from aim.metrics.interfaces import AIMMetricInterface
+
+# ----------------------------------------------------------------------------
+# Metadata
+# ----------------------------------------------------------------------------
+
+__author__ = "Amir Hossein Kargaran, Markku Laine, Thomas Langerak, Yuxi Zhu"
+__date__ = "2022-06-15"
+__email__ = "markku.laine@aalto.fi"
+__version__ = "2.0"
+
+
+# ----------------------------------------------------------------------------
+# Metric
+# ----------------------------------------------------------------------------
+
+
+class Metric(AIMMetricInterface):
+    """
+    Metric: Distinct values of Hue, Saturation, and Value.
+    """
+
+    # Private constants
+    _COLOR_REDUCTION_THRESHOLD_RATIO: float = (
+        0.001  # Threshold for minimum occuring of values: 0.1 %
+    )
+
+    # Public methods
+    @classmethod
+    def execute_metric(
+        cls,
+        gui_image: str,
+        gui_type: int = GUI_TYPE_DESKTOP,
+        gui_url: Optional[HttpUrl] = None,
+    ) -> Optional[List[Union[int, float, str]]]:
+        """
+        Execute the metric.
+
+        Args:
+            gui_image: GUI image (PNG) encoded in Base64
+
+        Kwargs:
+            gui_type: GUI type, desktop = 0 (default), mobile = 1
+            gui_url: GUI URL (defaults to None)
+
+        Returns:
+            Results (list of measures)
+            - Number of distinct Hue values (int, [1, 255])
+            - Number of distinct Saturation values (int, [1, 255])
+            - Number of distinct Value values (int, [1, 255])
+        """
+        # Create PIL image
+        img: Image.Image = Image.open(BytesIO(base64.b64decode(gui_image)))
+
+        # Convert image from ??? (e.g., RGBA) to HSV color space
+        # Note that all three values (Hue, Satuation, and Value) are between
+        # (0, 255): https://github.com/python-pillow/Pillow/issues/3650
+        img_hsv: Image.Image = img.convert("HSV")
+
+        # Calculate total number of image pixels
+        total_pixels: int = img_hsv.width * img_hsv.height
+
+        # Set color reduction threshold
+        color_reduction_threshold: int = int(
+            total_pixels * cls._COLOR_REDUCTION_THRESHOLD_RATIO
+        )
+
+        # Get HSV color histogram
+        hsv_color_histogram: List[Tuple[int, Tuple]] = img_hsv.getcolors(
+            maxcolors=total_pixels
+        )
+
+        # Create lists to store values of HSV, Hue, Saturation, and Value from the histogram
+        hue_values: List[int] = []
+        saturation_values: List[int] = []
+        value_values: List[int] = []
+
+        # Collect distinct values of Hue, Saturation, and Value after color reduction
+        for hist in list(hsv_color_histogram):
+            hist_count, hist_value = hist
+            if hist_count > color_reduction_threshold:
```

Comment:
m17 color reduction applied to hist_count, which is the number of HSV values.

Reply:
That's indeed tricky to judge based on the papers alone!

RGB:
- C4 in AVI'14: _"The number of distinct **RGB values**. Color variability was reduced: only **colors** covering more than five pixels were counted."_
- A3 in NordiCHI'14: _"Following the procedure from [21], we counted distinct **pixel RGB values** met more than two times"_

HSV:
- C1-C3 in AVI'14: _"The number of distinct **values of Hue, Saturation and Value.** Images were converted to the HSV color space. Color variability was reduced: only **values** covering more than 0.1% of image were counted."_
- A5-A7 in NordiCHI'14: _"counted distinct **Hue, Saturation and Value pixel values** coveting more than 0.1% of screenshot"_

Q1: Have you tried to find other implementations or literature on how to count unique colors/values in RGB and HSV? For example, [https://www.mathworks.com/matlabcentral/answers/77025-count-colours-in-hsv](https://www.mathworks.com/matlabcentral/answers/77025-count-colours-in-hsv)
Q2: Which of the two approaches makes more sense, if any?